File: src/test/java/com/raychenon/leetcode/ProductofArrayExceptSelfTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static org.junit.Assert.assertArrayEquals;

/**
 * User: raychenon
 * Date: 2019-08-18
 * https://leetcode.com/problems/product-of-array-except-self/
 */
public class ProductofArrayExceptSelfTest {

    private static ProductofArrayExceptSelf instance;

    @BeforeClass
    public static void setUp() {
        instance = new ProductofArrayExceptSelf();
    }

    @Test
    public void testProductOfArray() {
        int input[] = {1, 2, 3, 4};
        int output[] = {24, 12, 8, 6};
        assertArrays(output, input);
    }

    @Test
    public void testProductOfArraySizeZero() {
        // Given an array nums of n integers where n > 1
        int input[] = {};
        int output[] = {};
        assertArrays(output, input);
    }

    @Test
    public void testProductOfArray_when_containsZero() {
        int input[] = {1, 0};
        int output[] = {0, 1};
        assertArrays(output, input);
    }


    public void assertArrays(int[] output, int[] input) {
        assertArrayEquals(output, instance.productExceptSelf(input));
        assertArrayEquals(output, instance.productExceptSelfConstantSpace(input));
    }

}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/ProductofArrayExceptSelfTest.kt
package com.raychenon.kotlin.leetcode

import org.junit.Assert
import org.junit.Test

/**
 * User: raychenon
 * Date: 2019-08-19
 * https://leetcode.com/problems/product-of-array-except-self
 */
class ProductofArrayExceptSelfTest {

    @Test
    fun testProductArray() {
        val input = intArrayOf(1, 2, 3, 4)
        val output = intArrayOf(24, 12, 8, 6)

        assertArrays(input, output)
    }

    @Test
    fun testProductEmptyArray() {
        val input = intArrayOf()
        val output = intArrayOf()

        assertArrays(input, output)
    }

    @Test
    fun testProductArrayContainsZero() {
        val input = intArrayOf(1, 0)
        val output = intArrayOf(0, 1)

        assertArrays(input, output)
    }

    private fun assertArrays(actual: IntArray, expected: IntArray): Unit {
        Assert.assertArrayEquals(expected, ProductofArrayExceptSelf.productExceptSelf(actual))
        Assert.assertArrayEquals(expected, ProductofArrayExceptSelf.productExceptSelfConstantSpace(actual))
    }
}